src/main/java/seedu/taskell/logic/commands/UndoCommand.java

```diff
@@ -48,45 +72,64 @@ public CommandResult execute() {
             return new CommandResult(String.format(MESSAGE_NO_TASK_TO_UNDO));
         }
     }
+    
+    private static int getOffset(int index) {
+        return index - 1;
+    }
 
     private CommandResult undoDelete() {
         try {
-            model.addTask(mostRecentDeletedTask);
-            return new CommandResult(String.format(MESSAGE_ADD_TASK_SUCCESS, mostRecentDeletedTask));
+            model.addTask(commandHistory.getTask());
+            deleteCommandHistory();
+            return new CommandResult(String.format(MESSAGE_ADD_TASK_SUCCESS, commandHistory.getTask()));
         } catch (DuplicateTaskException e) {
             return new CommandResult(MESSAGE_DUPLICATE_TASK);
         }
     }
 
     private CommandResult undoAdd() {
         try {
-            model.deleteTask(mostRecentAddedTask);
+            model.deleteTask(commandHistory.getTask());
+            deleteCommandHistory();
         } catch (TaskNotFoundException e) {
             assert false : "The target task cannot be missing";
         }
-        return new CommandResult(String.format(MESSAGE_DELETE_TASK_SUCCESS, mostRecentAddedTask));
+        return new CommandResult(String.format(MESSAGE_DELETE_TASK_SUCCESS, commandHistory.getTask()));
     }
-    
-    public static void updateMostRecentCommand(String commandWord) {
-        mostRecentCommand = commandWord;
+
+    private void deleteCommandHistory() {
+        commandHistoryList.remove(commandHistory);
     }
     
-    public static void updateMostRecentAddedTask(Task task) {
-        mostRecentAddedTask = task;
+    public static void initializeCommandHistory() {
+        if (commandHistoryList==null) {
```

Comment:
Whitespace should be introduced.
`if (commandHistoryList == null) {`
![image](https://cloud.githubusercontent.com/assets/13817632/19630826/d6ab3660-99c4-11e6-9e6e-2f8b8f132b4d.png)